### ISSUE TITLE
Suggestions in .net7

### ISSUE THREAD

**Issue by narcis-ro** [meysamhadeli/booking-microservices#200]
Here are some suggestions from .net 7. Please take this as they are, suggestions, I'm not saying it's better or worse just some alternatives. Congratulations for this awesome sample!

- Use TypedResults and Multiple Results.
- In theory Hellang.Middleware.ProblemDetails is not needed anymore since the ProblemDetails are now easy configurable in .net 7
- Is SnowFlakIdGenerator since MassTransit already uses NewId with same purposes?
- Use required on props of records to simplify/omit some rules. 
- Question: Should booking service have any reference to a CurrentUser? Will this be called always in a context of a user? 

Lately we've been merging the the request, validator, handler in the same class. Since adding or removing the request will almost always involve changing the validator. F12 (Go to definition is magical). We do not use any auto discovery plain old container registration so navigation and find usages work. 

```
public static class Endpoint
{
    public static void MapRoute(IEndpointRouteBuilder route) => route.MapGet("/account/{id:int}", Handle);

    public record Request
    {
        public required int Id { get; init; }
    }

    public class Validator : AbstractValidator<Request>
    {
        public Validator() => RuleFor(x => x.Id).GreaterThan(2);
    }

    public record Response(string Desc);

    public static async Task<Results<Ok<Response>, NotFound>> Handle(
        [AsParameters] Request request,
        ILogger<Request> logger,
        ApiDbContext context,
        CancellationToken cancellationToken)
    {
        if (request.Id == 10)
        {
            return NotFound();
        }

        var entity = await context.Students.FirstOrDefaultAsync(cancellationToken: cancellationToken);

        return Ok(new Response($"everything went fine for: {request.Id}"));
    }
}
```
In case you want to do unit tests instead of acceptance tests of your endpoints you can inject Queries and Commands instead of db context so you can easily mock. 
We've tried TestContainers with server instance per session and each tests gets a new database + migration. It takes 10 sec to start sql server (azure sql edge) and 4ms to create a random new db with 2 tables. Running an acceptance test which: 1. generates new db. 2. creates the tables. 3. starts the api with test server. 4. makes a http get 5. make a query in db. 5. returns response. Takes 1 sec on average. This way ALL tests are insolated and run in parallel (even in the same (implicit) collection).
I'm planning a PoC for TestContainers with RabbitMq where only one instance of rabbitmq is started for a test session (per assembly fixutre) and all tests will use a separate vhost.

We've also dropped MediatR since we have very few truly cross concerns for it's pipeline and the number of request.
- OpenTelemetry adds logging to http and mass transit so tracebility is convered by that
- Caching at the api level. Since queries/command are never re-used between features they are tailed for each feature need. We add caching explicitly in the handler.
- Transactions when need are controlled at the handler level with transient dbcontext. A handler might create multiple transactions or call saves changes multiple times. In case of concurrency issues there are some chances to recover.
- Validation happens at the api/endpoint level at least the shallow one anything more complicated is handled directly in the handler

The principles we've been guiding ourselves:
- adhere to SOLID principles
- everything is an http api. Even cron jobs, background daemons. Use health-checks and probes
- vertical slices architecture
- invest a lot in identifying the correct bounded context
- duplication is far cheaper than wrong/premature abstractization
- we do not write applications to have tests, we write tests to build reliable applications. Refactoring should be done no matter how many tests need to be changed. Corollary prefer bdd, acceptance, integration and e2e tests which test business requirements instead of unit tests, especially when the MS are very small. Do sprinkle unit tests when business logic is complex.
- observability and traceability through OpenTelemetry are primordial
- easy to onboard new developers by self browsing since technical documentation is almost never written/updated.
  - drive developers into the pit of success with composable extensions framework. 
  - limit magic behaviors: auto mappers, auto registration reflection. Prefer source generators instead.
- YAGNI
  - We will not swap databases without effort
  - We will not change transport protocols from grpc/http to anything else without effort
  - We will not switch from an http api to a console app or daemon service without effort.

**Comment by meysamhadeli** [comment]
> 

Hi @narcis-ro, thanks for great suggestion, sorry because of my delay.

For better new managing, this suggestion it was better to `create separate issue for each topic`.

1- I prefer fail fast compared to functional programming (TypedResults , Multiple Results).
2- I don't familiar with this new feature in .net 7, if it has all feature in `Hellang.Middleware.ProblemDetails`, it's better we use it with new feature in .net 7. (feel free for creating a pull request)
3- I agree with you, I have plane for using NewId or Ulid in the future. (feel free for creating a pull request)
4- I agree with you, I have plane to change my props to required keywords. (feel free for creating a pull request)
5- We can get some information from token, it can be userId, username or claims in the all microservices.

> I will answer in the separate commands for other topics.

**Comment by meysamhadeli** [comment]
1- For endpoint, It is a matter of taste, and each one has its own preference. Your idea is attractive. But my way is common between developer I used this way. But I agree with you and in future I change it to this approach. (feel free for creating a pull request)
2- For test purpose, your approach is parallel, and we can run all test asynchronously, but it's a little slower. But I prefer to reset my database on the collection fixture, because it's a title faster, and it hast to be run sequential.
3- I use mediatr mostly for handling CQRS and use advantage of this approach to improve maintainability and some cross-cutting concern with some pipeline in my code.
4- I don't use open telemetry for logging, in my code, I use serilog for logging http calls. But for handling distributed tracing between microservices, I used open telemetry Instrumentations.
5- For caching, if we need to reuse a cache, it's a good idea to put it inner the handler. But for the simplest use case, my approach is faster and cleaner. (feel free for suggestion in the code)
6- For transaction section, I can't understand what you mean. If you have any suggestion, please send a pull request for clarifying.
7- Validation in endpoint level has a drawback. If we need to call another command, our query in our handler with using your approach, we will skip our validator. But with my approach, validations apply in business level. I use this article [REST API Data Validation](http://www.kamilgrzybek.com/design/rest-api-data-validation/) from kamil.

For principles, I will use your guidance. 
Please share to me any good resource by article or code.
Thanks.
